gvfs-helper.c

```diff
@@ -0,0 +1,2074 @@
+{
+	if (!gh__global.cache_server_is_initialized)
+		BUG("sub-command did not initialize cache-server vars");
+
+	if (!gh__global.cache_server_url)
+		return;
+
+	if (gh__global.cache_creds.username && *gh__global.cache_creds.username)
+		return;
+
+	/*
+	 * Get the main Git server creds so we can borrow the username
+	 * and password when we talk to the cache-server.
+	 */
+	lookup_main_creds();
+
+	// TODO we might not need to fully populate the creds for the
+	// TODO cache-server if we can completely disable the "store" and
+	// TODO "reject" events (in http.c) after the http request returns.
+	// TODO And if we can prevent a config scan.
+
+	credential_from_url(&gh__global.cache_creds, gh__global.cache_server_url);
+	gh__global.cache_creds.username = xstrdup(gh__global.main_creds.username);
+	gh__global.cache_creds.password = xstrdup(gh__global.main_creds.password);
+
+	// TODO consider deleting this message.
+	trace2_printf("cache_cred [host '%s'][user '%s']",
+		      gh__global.cache_creds.host,
+		      gh__global.cache_creds.username ?
+		      gh__global.cache_creds.username : "");
+}
+
+/*
+ * Flush and refresh the cache-server creds.  Because the cache-server
+ * does not do 401s (or manage creds), we have to reload the main Git
+ * server creds first.
+ *
+ * That is, we NEVER reject them directly because we never filled them.
+ */
+static void refresh_cache_server_creds(void)
+{
+	credential_clear(&gh__global.cache_creds);
+
+	refresh_main_creds();
+	synthesize_cache_server_creds();
+}
+
+/*
+ * We NEVER approve cache-server creds directly because we never directly
+ * filled them.  However, we should be able to infer that the main ones
+ * are valid and can approve them if necessary.
+ */
+static void approve_cache_server_creds(void)
+{
+	approve_main_creds();
+}
+
+/*
+ * Select the ODB directory where we will write objects that we
+ * download.  If no product-specific ODB is defined, use the
+ * local ODB (in ".git/objects").
+ */
+static void select_odb(void)
+{
+	const char *p_odb = the_repository->objects->odb->path;
+
+	strbuf_init(&gh__global.buf_odb_path, 0);
+
+	switch (gh__global.chosen_product_mode) {
+	default:                      /* should not happen */
+	case GH__PRODUCT_MODE__AUTO:  /* should not happen */
+	case GH__PRODUCT_MODE__NONE:
+		break;
+
+	case GH__PRODUCT_MODE__VFS:
+		if (gh__cmd_opts.vfs_odb_path)
+			p_odb = gh__cmd_opts.vfs_odb_path;
+		break;
+
+	case GH__PRODUCT_MODE__SCALAR:
+		if (gh__cmd_opts.scalar_odb_path)
+			p_odb = gh__cmd_opts.scalar_odb_path;
+		break;
+	}
+
+	strbuf_addstr(&gh__global.buf_odb_path, p_odb);
+}
+
+/*
+ * Create a tempfile to stream the packfile into.
+ *
+ * We create a tempfile in the chosen ODB directory and let CURL
+ * automatically stream data to the file.  If successful, we can
+ * later rename it to a proper .pack and run "git index-pack" on
+ * it to create the corresponding .idx file.
+ *
+ * TODO I would rather to just stream the packfile directly into
```

Comment:
This is an important TODO to handle in a follow-up. Let's create an issue for it.